### DISCUSSION TITLE
The global fixed random seed produces different training results every time

### DISCUSSION THREAD

**Discussion by Tyqfat** [pyg-team/pytorch_geometric#8716] [Q&A]
I encountered a problem when using pyg's GeneralConv (using edge features) for deep learning training. I have set the global fixed random seed, but during multiple runs of training, each time I get different training results. However, if I only use DNN or Transformer (without using edge features), the results of each training are the same. What is the cause of this? How should I solve this problem?
utils.py
```
import math
import os
import random
import time
import numpy as np
import torch

def set_random_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)  
    torch.cuda.manual_seed(seed)  
    torch.cuda.manual_seed_all(seed) 
    os.environ['PYTHONHASHSEED'] = str(seed)
    torch.backends.cudnn.deterministic = True  
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.enabled = True
```
network.py
```
import torch
import torch.nn as nn
from torch_geometric.nn import NNConv, GENConv, DeepGCNLayer, GeneralConv
import math
import torch.nn.functional as F


class GeneralConvNet(nn.Module):
    def __init__(self, args: dict, node_input_size, edge_input_size, node_output_size, edge_hidden_size,
                 node_hidden_size):
        super(GeneralConvNet, self).__init__()
        self.args = args
        self.node_output_size = node_output_size
        self.node_encoder = nn.Linear(node_input_size, node_hidden_size)
        self.edge_encoder = nn.Linear(edge_input_size, edge_hidden_size)
        self.general_conv_layers_1 = GeneralConv(in_channels=node_hidden_size, out_channels=node_output_size,
                                                 in_edge_channels=edge_hidden_size, aggr='mean', directed_msg=True,
                                                 heads=1)

    def forward(self, x, edge_index, edge_attr, num_graphs):
        x = self.node_encoder(x)
        edge_attr = self.edge_encoder(edge_attr)
        x = self.general_conv_layers_1(x, edge_index, edge_attr)
        x = x.view(num_graphs, -1, self.node_output_size)
        return x
```
main.py
```
if __name__ == '__main__':
    args = vars(get_params())
    set_random_seed(args['seed'])
    train_val_test(args)
```

**Comment by akihironitta** [comment]
Hi @Tyqfat, do you have your full code and env details to reproduce?

I haven't followed the status throughly, but looking at existing issues like https://github.com/pyg-team/pytorch_geometric/issues/2788#issuecomment-870502307, aggregations are not deterministic unless you use `SparseTensor`.

**Reply by Tyqfat** [reply]
Thanks for the quick reply. Since my project code is too large to render all of it, I will look at the answer in https://github.com/pyg-team/pytorch_geometric/issues/2788#issuecomment-870502307.

**Reply by akihironitta** [reply]
FYI, I've labelled all issues and discussions that are related to non-deterministic behaviours 😉
- https://github.com/pyg-team/pytorch_geometric/issues?q=is%3Aissue+sort%3Aupdated-desc+label%3Adeterminism
- https://github.com/pyg-team/pytorch_geometric/discussions?discussions_q=label%3Adeterminism

**Reply by Tyqfat** [reply]
Thanks for your help.

**Reply by Tyqfat** [reply]
Hi, @akihironitta. I am using `SparseTensor`, and the model produces consistent results during each training session on the CPU; however, it still exhibits non-determinism when running on the GPU. Could you please advise what might be causing this issue and how to resolve it?

network.py
```
class GeneralConvNet(nn.Module):
    def __init__(self, args: dict, node_input_size, edge_input_size, node_output_size, edge_hidden_size,
                 node_hidden_size):
        super(GeneralConvNet, self).__init__()
        self.args = args
        self.node_output_size = node_output_size
        self.node_encoder = nn.Linear(node_input_size, node_hidden_size)
        self.edge_encoder = nn.Linear(edge_input_size, edge_hidden_size)
        self.general_conv_layers_1 = GeneralConv(in_channels=node_hidden_size, out_channels=node_output_size,
                                                 in_edge_channels=edge_hidden_size, aggr='mean', directed_msg=True,
                                                 heads=1)

    def forward(self, x, edge_index, edge_attr, num_graphs):
        x = self.node_encoder(x)
        edge_attr = self.edge_encoder(edge_attr)
        adj = SparseTensor(row=edge_index[0], col=edge_index[1], value=edge_attr, sparse_sizes=(x.size(0), x.size(0)))
        # x = self.general_conv_layers_1(x, edge_index, edge_attr)
        x = self.general_conv_layers_1(x, adj.t())
        x = x.view(num_graphs, -1, self.node_output_size)
        return x
```
![image](https://github.com/pyg-team/pytorch_geometric/assets/100210376/203103e4-985e-4c29-9ecf-a7688cd81223)

**Reply by rusty1s** [reply]
How do you measure the non-determinism? Did you set `seed_everything()`?

**Reply by Tyqfat** [reply]
utils.py
```
import math
import os
import random
import time
import numpy as np
import torch

def set_random_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)  
    torch.cuda.manual_seed(seed)  
    torch.cuda.manual_seed_all(seed) 
    os.environ['PYTHONHASHSEED'] = str(seed)
    torch.backends.cudnn.deterministic = True  
    torch.backends.cudnn.benchmark = False
    torch.backends.cudnn.enabled = True
```
network.py
```
import torch
import torch.nn as nn
from torch_geometric.nn import NNConv, GENConv, DeepGCNLayer, GeneralConv
import math
import torch.nn.functional as F


class GeneralConvNet(nn.Module):
    def __init__(self, args: dict, node_input_size, edge_input_size, node_output_size, edge_hidden_size,
                 node_hidden_size):
        super(GeneralConvNet, self).__init__()
        self.args = args
        self.node_output_size = node_output_size
        self.node_encoder = nn.Linear(node_input_size, node_hidden_size)
        self.edge_encoder = nn.Linear(edge_input_size, edge_hidden_size)
        self.general_conv_layers_1 = GeneralConv(in_channels=node_hidden_size, out_channels=node_output_size,
                                                 in_edge_channels=edge_hidden_size, aggr='mean', directed_msg=True,
                                                 heads=1)

    def forward(self, x, edge_index, edge_attr, num_graphs):
        x = self.node_encoder(x)
        edge_attr = self.edge_encoder(edge_attr)
        adj = SparseTensor(row=edge_index[0], col=edge_index[1], value=edge_attr, sparse_sizes=(x.size(0), x.size(0)))
        # x = self.general_conv_layers_1(x, edge_index, edge_attr)
        x = self.general_conv_layers_1(x, adj.t())
        x = x.view(num_graphs, -1, self.node_output_size)
        return x
```
main.py
```
if __name__ == '__main__':
    args = vars(get_params())
    set_random_seed(args['seed'])
    train_val_test(args)
```

I have set `set_random_seed`, Here are the results of two experiments.
run1:
```
device: cuda
All dataset size: 22452
Train dataset size: 17961
Val dataset size: 2245
Test dataset size: 2246
-- Epoch: 01 	Training Loss: 0.341798 	Validation Loss: 0.330329
Validation loss decreased (inf --> 0.330329).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 01 time: 8.95 seconds 	Total time: 00h:00m:8.95s	 Total seconds:8.95s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 02 	Training Loss: 0.331268 	Validation Loss: 0.330106
Validation loss decreased (0.330329 --> 0.330106).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 02 time: 2.88 seconds 	Total time: 00h:00m:11.84s	 Total seconds:11.84s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 03 	Training Loss: 0.330841 	Validation Loss: 0.329742
Validation loss decreased (0.330106 --> 0.329742).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 03 time: 2.87 seconds 	Total time: 00h:00m:14.71s	 Total seconds:14.71s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 04 	Training Loss: 0.330597 	Validation Loss: 0.329445
Validation loss decreased (0.329742 --> 0.329445).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 04 time: 2.87 seconds 	Total time: 00h:00m:17.58s	 Total seconds:17.58s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 05 	Training Loss: 0.330440 	Validation Loss: 0.329298
Validation loss decreased (0.329445 --> 0.329298).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 05 time: 2.89 seconds 	Total time: 00h:00m:20.46s	 Total seconds:20.46s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 06 	Training Loss: 0.330264 	Validation Loss: 0.329359
EarlyStopping counter: 1 out of 20
-- Learning Rate: 0.001
-- Epoch: 06 time: 2.93 seconds 	Total time: 00h:00m:23.40s	 Total seconds:23.40s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 07 	Training Loss: 0.329819 	Validation Loss: 0.323571
Validation loss decreased (0.329298 --> 0.323571).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 07 time: 2.98 seconds 	Total time: 00h:00m:26.38s	 Total seconds:26.38s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 08 	Training Loss: 0.319504 	Validation Loss: 0.314181
Validation loss decreased (0.323571 --> 0.314181).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 08 time: 2.91 seconds 	Total time: 00h:00m:29.29s	 Total seconds:29.29s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 09 	Training Loss: 0.309428 	Validation Loss: 0.304185
Validation loss decreased (0.314181 --> 0.304185).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 09 time: 2.87 seconds 	Total time: 00h:00m:32.16s	 Total seconds:32.16s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 10 	Training Loss: 0.303217 	Validation Loss: 0.296678
Validation loss decreased (0.304185 --> 0.296678).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 10 time: 2.88 seconds 	Total time: 00h:00m:35.03s	 Total seconds:35.03s
----------------------------------------------------------------------------------------------------------------------------------
Best valid loss: 0.296678
loading model
Test generator_classification shape: (2246, 24, 10)
-- generator_classification test Accuracy: 0.8285
-- test time: 0.26 seconds
```
run2:
```
device: cuda
All dataset size: 22452
Train dataset size: 17961
Val dataset size: 2245
Test dataset size: 2246
-- Epoch: 01 	Training Loss: 0.341798 	Validation Loss: 0.330328
Validation loss decreased (inf --> 0.330328).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 01 time: 9.90 seconds 	Total time: 00h:00m:9.90s	 Total seconds:9.90s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 02 	Training Loss: 0.331267 	Validation Loss: 0.330108
Validation loss decreased (0.330328 --> 0.330108).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 02 time: 3.01 seconds 	Total time: 00h:00m:12.91s	 Total seconds:12.91s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 03 	Training Loss: 0.330842 	Validation Loss: 0.329739
Validation loss decreased (0.330108 --> 0.329739).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 03 time: 3.02 seconds 	Total time: 00h:00m:15.93s	 Total seconds:15.93s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 04 	Training Loss: 0.330598 	Validation Loss: 0.329444
Validation loss decreased (0.329739 --> 0.329444).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 04 time: 3.05 seconds 	Total time: 00h:00m:18.98s	 Total seconds:18.98s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 05 	Training Loss: 0.330441 	Validation Loss: 0.329299
Validation loss decreased (0.329444 --> 0.329299).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 05 time: 3.04 seconds 	Total time: 00h:00m:22.03s	 Total seconds:22.03s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 06 	Training Loss: 0.330265 	Validation Loss: 0.329252
Validation loss decreased (0.329299 --> 0.329252).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 06 time: 3.10 seconds 	Total time: 00h:00m:25.13s	 Total seconds:25.13s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 07 	Training Loss: 0.328984 	Validation Loss: 0.328967
Validation loss decreased (0.329252 --> 0.328967).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 07 time: 3.16 seconds 	Total time: 00h:00m:28.29s	 Total seconds:28.29s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 08 	Training Loss: 0.319613 	Validation Loss: 0.316526
Validation loss decreased (0.328967 --> 0.316526).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 08 time: 3.06 seconds 	Total time: 00h:00m:31.34s	 Total seconds:31.34s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 09 	Training Loss: 0.309380 	Validation Loss: 0.300711
Validation loss decreased (0.316526 --> 0.300711).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 09 time: 3.05 seconds 	Total time: 00h:00m:34.39s	 Total seconds:34.39s
----------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------
-- Epoch: 10 	Training Loss: 0.301834 	Validation Loss: 0.295632
Validation loss decreased (0.300711 --> 0.295632).  Saving model ...
-- Learning Rate: 0.001
-- Epoch: 10 time: 3.05 seconds 	Total time: 00h:00m:37.44s	 Total seconds:37.44s
----------------------------------------------------------------------------------------------------------------------------------
Best valid loss: 0.295632
loading model
Test generator_classification shape: (2246, 24, 10)
-- generator_classification test Accuracy: 0.8295
-- test time: 0.25 seconds
```

**Reply by rusty1s** [reply]
Got it. This seems to be an issue with `GeneralConv`. Can you confirm that you are not seeing this when using, e.g., `SAGEConv`?

**Reply by Tyqfat** [reply]
I have found the problem. It is the uncertain results caused by the adam optimizer. When I use the SGD optimizer with momentum, the results of every experiment will be the same.

**Reply by Tyqfat** [reply]
Why does using the adam optimizer lead to uncertain results? I'm confused.

**Reply by rusty1s** [reply]
Are you sure it's a problem with `Adam`? That would be new to me.

**Reply by Tyqfat** [reply]
After re-inspecting the code, I apologize for my previous understanding; the `non-deterministic` behavior was not caused by the `Adam` optimizer, but rather stemmed from the usage of `CUDA`. Upon adding the following statement within the `set_random_seed` function, consistent results are obtained across experiments regardless of whether `GeneralConv` employs `SparseTensor` or not.
```
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8" # my cuda version = 11.8 
torch.use_deterministic_algorithms(True)
```
utils.py
new `set_random_seed` function
```
def set_random_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)  
    torch.cuda.manual_seed(seed)  
    torch.cuda.manual_seed_all(seed)  
    os.environ['PYTHONHASHSEED'] = str(seed)
    torch.backends.cudnn.enabled = True   
    torch.backends.cudnn.deterministic = True  
    torch.backends.cudnn.benchmark = False 
    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
    print(f'CUBLAS_WORKSPACE_CONFIG: {os.environ["CUBLAS_WORKSPACE_CONFIG"]}') # You can print it out to see if the value has been set successfully
    torch.use_deterministic_algorithms(True) 
```

**Reply by rusty1s** [reply]
Ah, that's good to know. Do you mind contributing your seed function to `seed_everything` in PyG?

**Reply by Tyqfat** [reply]
Of course, I'd love to!(●'◡'●)
